### PR TITLE
[REVIEW] Improve Wavelets functions performance

### LINKED ISSUE #223: [FEA] Reevaluate Wavelets module for performance improvements
The wavelets module need to be reevaluated to find areas of further performance improvements.

### LINKED ISSUE #234: [BUG] fftconvolve error when specifying axes
**Describe the bug**
I experience the following error when using `cusignal.convolution.fftconvolve`  with cupy arrays in place of `scipy.signal.fftconvolve` with numpy arrays:
```
/usr/local/lib/python3.6/site-packages/cupy/_indexing/indexing.py in take(a, indices, axis, out)
     27     """
     28     # TODO(okuta): check type
---> 29     return a.take(indices, axis, out)
     30 
     31 

AttributeError: 'tuple' object has no attribute 'take'
```
The error originates when setting the `axes` input parameter to apply convolution over only one or a set of axes. Seems to work fine when leaving it empty (and so applying the convolution over all the axes)

 
**Steps/Code to reproduce bug**
Working code
```
#scipy
x = np.zeros((20, 10))
x[10, 5] = 1
h = np.ones((4, 1))

y = sp.signal.fftconvolve(x,h, mode='same', axes=0)
```
Broken code
```
#cusignal
x = cp.zeros((20, 10))
x[10, 5] = 1
h = cp.ones((4, 1))

y = cusignal.convolution.fftconvolve(x, h, mode='same', axes=0)
```

**Expected behavior**
Same behaviour as scipy fftconvolve (or remove axes?)

**Environment details (please complete the following information):**
 - Environment location: Colab
 - Method of cuSignal install: conda

**Additional context**
Python version: 3.6
Cupy version: 8.0.0rc1
Cusignal version: 0.14.0

### PR DESCRIPTION
**Pull request by mnicely**
Closes #223 
Closes #234 

This PR improves performance for all wavelets functions.

### branch-0.16
```bash
-------------- benchmark 'CWT': 4 tests --------------
Name (time in ms, mem in bytes)         Mean          
------------------------------------------------------
test_cwt_gpu[31-16384]               72.6608 (5.04)   
test_cwt_gpu[127-16384]             310.6303 (21.54)  
------------------------------------------------------

---------- benchmark 'CWTComplex': 4 tests -----------
Name (time in ms, mem in bytes)         Mean          
------------------------------------------------------
test_cwt_complex_gpu[31-16384]       71.0334 (3.05)   
test_cwt_complex_gpu[127-16384]     300.4632 (12.91)  
------------------------------------------------------

------------ benchmark 'Morlet': 2 tests -------------
Name (time in us, mem in bytes)         Mean          
------------------------------------------------------
test_morlet_gpu[16384]              166.5411 (1.0)    
------------------------------------------------------

------------ benchmark 'Ricker': 4 tests -------------
Name (time in us, mem in bytes)         Mean          
------------------------------------------------------
test_ricker_gpu[1000-16384]         216.0455 (1.53)   
test_ricker_gpu[10-16384]           216.7407 (1.54)   
------------------------------------------------------

```

### This PR
```
-------------- benchmark 'CWT': 4 tests -------------
Name (time in ms, mem in bytes)        Mean          
-----------------------------------------------------
test_cwt_gpu[31-16384]              16.6778 (1.17)   
test_cwt_gpu[127-16384]             67.9339 (4.77)   
-----------------------------------------------------

---------- benchmark 'CWTComplex': 4 tests -----------
Name (time in ms, mem in bytes)         Mean          
------------------------------------------------------
test_cwt_complex_gpu[31-16384]       14.6283 (1.0)    
test_cwt_complex_gpu[127-16384]      61.4721 (4.20)   
------------------------------------------------------

------------ benchmark 'Morlet': 2 tests -------------
Name (time in us, mem in bytes)         Mean          
------------------------------------------------------
test_morlet_gpu[16384]               28.6952 (1.0)    
------------------------------------------------------

--------------- benchmark 'Qmf': 2 tests ---------------
Name (time in us, mem in bytes)           Mean          
--------------------------------------------------------
test_qmf_gpu[16384]                    13.3289 (1.0)    
--------------------------------------------------------

------------ benchmark 'Ricker': 4 tests -------------
Name (time in us, mem in bytes)         Mean          
------------------------------------------------------
test_ricker_gpu[1000-16384]          23.1164 (1.0)    
test_ricker_gpu[10-16384]            23.1946 (1.00)   
------------------------------------------------------
```

Improvements in this PR are made by move many arithmetic calculation from CuPy functions to a CuPy Elementwise kernel. This minimize kernel launch overhead. 

Replace
```python
def morlet(M, w=5.0, s=1.0, complete=True):
    x = cp.linspace(-s * 2 * cp.pi, s * 2 * cp.pi, M)
    output = cp.exp(1j * w * x)

    if complete:
        output -= cp.exp(-0.5 * (w**2))

    output *= cp.exp(-0.5 * (x**2)) * cp.pi**(-0.25)

    return output
```

with
```
_morlet_kernel = cp.ElementwiseKernel(
    "float64 delta, float64 start, float64 w, float64 pi, bool complete",
    "T output",
    """
    double x = start + delta * i;
    T temp = T(0, w * x);
    temp = exp(temp);
    if (complete) {
        temp -= exp( -0.5 * (w * w) );
    }
    output = temp * exp( -0.5 * (x * x)) * pow(pi, -0.25)
    """,
    "_morlet_kernel",
)

def morlet(M, w=5.0, s=1.0, complete=True):
    output = cp.empty(M, dtype=cp.complex128)
    end = s * 2 * np.pi
    start = -s * 2 * np.pi
    delta = (end - start) / (M - 1)

    _morlet_kernel(delta, start, w, np.pi, complete, output, size=M)

    return output
```

For functions like `cwt`, improvements are made by moving operations like scalar arithmetic and/or comparison back to the CPU using NumPy. Along with add a FFT cache dictionary to remove redundant FFT plan creations!

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by awthomp** [comment]
Wow -- so we're getting ~4-5x speedup here?

**Comment by mnicely** [comment]
> Wow -- so we're getting ~4-5x speedup here?

Yes sir! :smile:

**Comment by awthomp** [comment]
> > Wow -- so we're getting ~4-5x speedup here?
> 
> Yes sir! 😄 Or more!

I'd totally include this in your GTC talk, and I'll warn about it in the community. While drop-in lib replacements are amazing, you gotta make sure that operation isn't faster on CPU, especially if you're just calculating things like FFT sizes, number of samples to overlap, and other general parameters.

File: python/cusignal/convolution/convolve.py [hunk 1]
```diff
@@ -299,19 +304,38 @@ def fftconvolve(in1, in2, mode="full", axes=None):
...
+        else:
+            irplan = _cupy_fft_cache[
+                (str(fshape), str(axes), "C2R")
+            ] = fftpack.get_fft_plan(in1, fshape, axes=axes, value_type="C2R")
+
+        with rplan:
+            sp1 = cp.fft.rfftn(in1, fshape, axes=axes)
+            sp2 = cp.fft.rfftn(in2, fshape, axes=axes)
+        with irplan:
+            ret = cp.fft.irfftn(sp1 * sp2, fshape, axes=axes)[fslice].copy()
```
Comment (awthomp) [inline python/cusignal/convolution/convolve.py:325]: Out of curiosity, how much of the ~5x perf gain is from cached FFTs?

Comment (mnicely) [reply python/cusignal/convolution/convolve.py:325]: A bulk of the time for `cwt`... Maybe 50%

**Review by awthomp — APPROVED** [review]
Great PR. Before merging, can you please summarize your best practices for improving performance so that other folks can contribute to doing the same types of work to the other FEA requests?

File: python/cusignal/convolution/convolve.py (hunk 1 above)
Comment (mnicely) [reply python/cusignal/convolution/convolve.py:325]: We also see improvements for all `fftconvolve` usage. But I'll show that in a different PR.

**Comment by mnicely** [comment]
@awthomp Currently, this module matches the usage of Scipy Signal wavelet functions. Meaning output for all functions is `float64` or `complex128`.

Did we want to give cuSignal users the option to input/output single precision?

We could add a function parameter `dtype=cp.float64` that they can override.

**Comment by mnicely** [comment]
> Great PR. Before merging, can you please summarize your best practices for improving performance so that other folks can contribute to doing the same types of work to the other FEA requests?

Updated original comment

**Comment by awthomp** [comment]
> @awthomp Currently, this module matches the usage of Scipy Signal wavelet functions. Meaning output for all functions is `float64` or `complex128`.
> 
> Did we want to give cuSignal users the option to input/output single precision?
> 
> We could add a function parameter `dtype=cp.float64` that they can override.

Let's add single precision support but do it in a separate PR. I'm good to merge this! LGTM

**Comment by mnicely** [comment]
> > @awthomp Currently, this module matches the usage of Scipy Signal wavelet functions. Meaning output for all functions is `float64` or `complex128`.
> > Did we want to give cuSignal users the option to input/output single precision?
> > We could add a function parameter `dtype=cp.float64` that they can override.
> 
> Let's add single precision support but do it in a separate PR. I'm good to merge this! LGTM

Same for Windows module?

**Comment by awthomp** [comment]
Yeah, let's see what the community wants before expanding support -- as trivial as it may be to code.

**Pull request merged**